### PR TITLE
Quantized leaky relu

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33004 Quantized leaky relu**

Differential Revision: [D19740193](https://our.internmc.facebook.com/intern/diff/D19740193)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -193,6 +193,51 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
...
+  float o_scale = out.q_scale();
+  float o_inv_scale = 1.0 / o_scale;
+
+  float negval = negval_.to<float>();
+
+  AT_DISPATCH_QINT_TYPES(out.scalar_type(), "leaky_qrelu", [&] {
+    using Vec = Vec256<float>;  // Naïve implementation uses dequant/quant loop.
+    using qVec = Vec256<scalar_t>;
+    Vec zero_vec = Vec((float)(0.0f));
+    Vec one_vec = Vec((float)(1.0f));
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Why do we need explicit casts on float literals?

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 2]
```diff
@@ -193,6 +193,51 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
...
 }
 
+static void leaky_qrelu_out_kernel(Tensor& out, const Tensor& qx,
+                                   Scalar negval_) {
+  int64_t i_zp = qx.q_zero_point();
+  float i_scale = qx.q_scale();
+
+  int64_t o_zp = out.q_zero_point();
+  float o_scale = out.q_scale();
+  float o_inv_scale = 1.0 / o_scale;
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: 1.0f

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 3]
```diff
@@ -193,6 +193,51 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
...
+          auto value_dx = at::dequantize_val(i_scale, i_zp, value_qx);
+          auto value_dy = value_dx > 0 ? value_dx : value_dx * negval;
+          return at::quantize_val<scalar_t>(o_scale, o_zp, value_dy);
+        },
+        [&](qVec qx_vec) -> qVec {
+          auto dx_vec_vec = qx_vec.dequantize(i_scale_vec, i_zp_vec,
+                                              i_scale_zp_neg_premul_vec);
+          for (int idx = 0; idx < dx_vec_vec.size(); ++idx) {
+            const auto dx_vec = dx_vec_vec[idx];
+            const auto r = Vec::blendv(negval_vec, one_vec, dx_vec > zero_vec);
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Can we either name this something more semantic than `r` or place a mathematical formula in a comment that people can refer to to see what `r` is?

File: test/test_quantized.py [hunk 4]
```diff
@@ -148,6 +148,27 @@ def test_qrelu6(self, X):
                 self.assertEqual(qY, qY_hat,
                                  message="{} relu failed".format(name))
 
+    """Tests the correctness of the quantized::relu op."""
+    @given(X=hu.tensor(shapes=hu.array_shapes(1, 5, 1, 5),
+                       qparams=hu.qparams()))
+    def test_qrelu_leaky(self, X):
+        X, (scale, zero_point, torch_type) = X
+        alpha = 0.1
```
Comment (jamesr66a) [inline test/test_quantized.py]: Should we test different values for this? Though i like the choice of `0.1` for its pathological floating point representation :)

**Review by jamesr66a — APPROVED** [review]
LG, small comments

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 1 above)
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: We don't -- just a habit :)

**Pull request closed without merging**